File: sorento_crm_backend/app/services/revision_fence.py
"""The revision fence: no office action may land against a superseded version.

UAC ``portal-submission-revisions`` section C-bis (CB1-CB4).

Once a contact revision voids the stage an office user was working, that user's
open tab is holding a version that no longer exists. Voiding the tracker alone
does not stop them: the stock-inquiry respond path stamps ``last_responded_*``
and messages the contact from any status, and every lifecycle action (approve,
reject, process, close, void) would otherwise apply cleanly to the new revision
while having been decided against the old one (CB2).

So every office write on a revisable form carries the ``revision_no`` the client
was looking at, in the ``X-Revision-No`` request header, and this refuses the
write with **409** when it no longer matches.

Three deliberate properties:

* **One shared dependency** (CB3). A new revisable form type inherits the fence
  by adding a target below plus its config row, never by re-deriving the check
  per endpoint.
* **The header is optional.** A request that does not carry it is not fenced.
  That is what keeps every existing integration principal working (n8n, the MCP
  server, the external API), none of which has any notion of a revision. The
  fence protects the surface that CAN send it - the office UI - and the portal
  revise path has its own, mandatory, ``expected_revision_no`` guard (UAC C5).
* **Revisability is config-driven** (CB4). A type whose
  ``portal_revision_configs`` row is disabled - complaint today - costs one
  cached-free config read and nothing else; the row is never loaded.

NOT fenced, on purpose: ``POST /{entity}/{id}/conversation/send-message``. That
is the plain chat send, it never mutates the entity, and AC O2 keeps messaging a
contact working at any status and any revision. The fence guards entity writes,
not the conversation.

Ordering note: FastAPI solves route-level dependencies BEFORE the handler's own,
so on a permission-gated route a stale header yields 409 rather than 403. The
fence authenticates (below) so this is never reachable anonymously; beyond that,
telling an authenticated user their view is stale before telling them they lack
the permission is a harmless ordering, and the write still never happens.
"""
from __future__ import annotations

import logging
import uuid
from dataclasses import dataclass
from typing import Any, Callable, Optional

from fastapi import Depends, Header, Request
from sqlalchemy.orm import Session

from app.database import get_db
from app.dependencies import get_current_user_or_api_key
from app.services.error_handler import handle_conflict

logger = logging.getLogger(__name__)

#: The header an office client sends with the revision it was viewing.
REVISION_HEADER = "X-Revision-No"


@dataclass(frozen=True)
class _FenceTarget:
    """One fenced route family: how to load the row and what type it is.

    ``entity_types`` is every ``source_entity_type`` the target can resolve to,
    used to skip the row load entirely when none of them is revisable.
    ``resolve_entity_type`` picks the exact one off the loaded row - purchase
    request and sponsorship form share a table, discriminated by
    ``request_type``.
    """

    load_model: Callable[[], Any]
    entity_types: tuple[str, ...]
    resolve_entity_type: Callable[[Any], str]


def _stock_inquiry_model() -> Any:
    from app.models.procurement import StockInquiry

    return StockInquiry


def _purchase_request_model() -> Any:
    from app.models.procurement import PurchaseRequestHeader

    return PurchaseRequestHeader


def _complaint_model() -> Any:
    from app.models.complaints import Complaint

    return Complaint


STOCK_INQUIRY = "stock_inquiry"
PURCHASE_REQUEST = "purchase_request"
COMPLAINT = "complaint"

_TARGETS: dict[str, _FenceTarget] = {
    STOCK_INQUIRY: _FenceTarget(
        load_model=_stock_inquiry_model,
        entity_types=("stock_inquiry",),
        resolve_entity_type=lambda row: "stock_inquiry",
    ),
    # One route family, two revisable types. Both are enabled from day one, so
    # the cheap pre-check passes and the row decides which config applies.
    PURCHASE_REQUEST: _FenceTarget(
        load_model=_purchase_request_model,
        entity_types=("purchase_request", "sponsorship_form"),
        resolve_entity_type=lambda row: (
            (getattr(row, "request_type", None) or "purchase_request").strip().lower()
            or "purchase_request"
        ),
    ),
    # Wired but inert: the complaint config row ships disabled (UAC A5), so this
    # short-circuits before touching the database. Flipping the checkbox arms it
    # with no code change - which is the point of CB4.
    COMPLAINT: _FenceTarget(
        load_model=_complaint_model,
        entity_types=("complaint",),
        resolve_entity_type=lambda row: "complaint",
    ),
}

# How the refusal names the document. Same vocabulary the contact-facing
# revision copy uses, so the two sides of the feature read alike.
_LABELS: dict[str, str] = {
    "stock_inquiry": "stock inquiry",
    "purchase_request": "purchase request",
    "sponsorship_form": "sponsorship form",
    "complaint": "complaint",
}


def _is_uuid(value: Any) -> bool:
    try:
        uuid.UUID(str(value))
        return True
    except (ValueError, TypeError, AttributeError):
        return False


def _revisable_types(db: Session, candidates: tuple[str, ...]) -> set[str]:
    """Which of ``candidates`` currently has an ENABLED revision config.

    Fails closed on content (a missing row means disabled, UAC A3) and open on
    infrastructure: if the config table is not there at all, the fence is simply
    absent rather than 500-ing every office write.
    """
    try:
        from app.models.portal import PortalRevisionConfig

        rows = (
            db.query(PortalRevisionConfig.source_entity_type, PortalRevisionConfig.is_enabled)
            .filter(PortalRevisionConfig.source_entity_type.in_(candidates))
            .all()
        )
    except Exception:  # noqa: BLE001 - partial schema; never block a write on the fence
        db.rollback()
        logger.warning("Revision fence: config lookup failed for %s", candidates, exc_info=True)
        return set()
    return {str(entity_type) for entity_type, enabled in rows if enabled}


def assert_revision_current(
    db: Session,
    target: str,
    entity_id: Any,
    expected_revision_no: Optional[int],
) -> None:
    """Refuse a write aimed at a superseded revision (CB1). 409, or nothing.

    Silent no-op when: the client sent no expectation, the id is not a UUID (the
    route's own validation owns that error), the row is gone (the route's 404
    owns that), or the type is not currently revisable.
    """
    if expected_revision_no is None:
        return
    spec = _TARGETS.get(target)
    if spec is None or not entity_id or not _is_uuid(entity_id):
        return

    enabled = _revisable_types(db, spec.entity_types)
    if not enabled:
        return

    model = spec.load_model()
    row = db.query(model).filter(model.id == str(entity_id)).first()
    if row is None:
        return
    if spec.resolve_entity_type(row) not in enabled:
        return

    current = int(getattr(row, "revision_no", 0) or 0)
    if int(expected_revision_no) == current:
        return

    label = _LABELS.get(spec.resolve_entity_type(row), "record")
    # A mismatch does not always mean "a revision happened". When `current` is 0 the
    # record has never been revised, so the client sent a number that was never real
    # (a stale tab from before a reset, a hand-built request, a bug). Saying "was
    # revised ... reload to see revision 0" in that case is self-contradictory and
    # tells the reader nothing they can act on, so name only what is certain.
    if current == 0:
        raise handle_conflict(
            f"This {label} has changed since you opened it. Reload and try again."
        )
    raise handle_conflict(
        f"This {label} was revised while you were working on it. "
        f"Reload to see revision {current}."
    )


def require_current_revision(target: str, id_param: str) -> Callable:
    """FastAPI dependency for one fenced route family.

    ``target`` keys :data:`_TARGETS`; ``id_param`` is the path parameter holding
    the entity id (``inquiry_id`` / ``request_id`` / ``complaint_id``), read off
    the request rather than declared, because the name differs per router and
    the whole point is one dependency.

    Mount with ``dependencies=[Depends(require_current_revision(...))]`` so the
    check runs before the handler and no write happens on a refusal.
    """

    def _fence(
        request: Request,
        x_revision_no: Optional[int] = Header(
            default=None,
            alias=REVISION_HEADER,
            description=(
                "The revision_no the client was viewing. When present and stale, "
                "the write is refused with 409."
            ),
        ),
        # Route-level dependencies are solved BEFORE the handler's own, so without
        # this an anonymous caller could probe a record's revision through the 409.
        # Authenticating here keeps the fence behind the same door as the write.
        _principal: dict = Depends(get_current_user_or_api_key),
        db: Session = Depends(get_db),
    ) -> None:
        if x_revision_no is None:
            return
        assert_revision_current(
            db, target, request.path_params.get(id_param), x_revision_no
        )

    return _fence

File: sorento_crm_frontend/lib/revision-fence.ts
/**
 * The office side of the revision fence (UAC `portal-submission-revisions`, C-bis).
 *
 * The backend mounts `require_current_revision(...)` on every office write that
 * mutates a revisable form: 34 routes across stock inquiry, purchase request /
 * sponsorship form and complaint. It reads `X-Revision-No`, compares it to the
 * row's current `revision_no`, and refuses a stale write with 409 plus one plain
 * sentence.
 *
 * It is 34 routes and not "everything under a record": the chat send, the view
 * link, the PDF export and the assignee sync are all deliberately left open, so
 * this file mirrors the fenced set rather than inferring it (`FENCED_COLLECTIONS`).
 *
 * **An absent header means unfenced, by design** - that is what keeps n8n, the
 * MCP server and the external API working, none of which knows what a revision
 * is. So a fence nobody sends the header to protects nothing: the office UI is
 * the one surface the fence exists for, and it is the one surface that has to
 * send it.
 *
 * ## Why this lives under `apiFetch` and not in the 34 call sites
 *
 * Sending the header is a property of the REQUEST URL, not of any one feature.
 * Threading a `revisionNo` argument through 34 service functions, their hooks
 * and their components would be 34 chances to forget one, and every new fenced
 * route would silently start out unfenced. `apiFetch` is the single choke point
 * every one of those writes already passes through, and it already carries an
 * interceptor of exactly this shape (`_attachImpersonationHeader`). So:
 *
 * * **Capture** - a GET of a fenced list or detail harvests `{id, revision_no}`
 *   off the response into a module-level registry. That is literally "the
 *   revision the user was looking at": it is the number the screen rendered.
 *   Capturing from the list too means a row action taken without ever opening
 *   the detail page is fenced as well.
 * * **Send** - a non-GET to one of the routes the backend actually fences (see
 *   `FENCED_COLLECTIONS`, an allowlist rather than "anything under a record")
 *   carries the remembered revision. Nothing remembered -> no header -> unfenced,
 *   which is the honest answer when we genuinely do not know what was on screen.
 * * **Refuse** - a 409 on a request WE fenced is normalized so the server's
 *   sentence reaches the caller whichever way it reads the body, and the record
 *   is refreshed so the user lands on the new revision instead of staring at the
 *   superseded one behind a toast.
 *
 * The registry keeps the HIGHEST revision seen per id. `revision_no` only ever
 * increments, so a stale list response arriving after a fresh detail response
 * cannot walk the number backwards and manufacture a 409.
 */

import { extractApiError } from '@/lib/api-client';

/** The header the backend's `revision_fence.py` reads. Same spelling, one place. */
export const REVISION_HEADER = 'X-Revision-No';

const UUID_PATTERN =
  '[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}';

/**
 * The fenced routes, one entry per fenced router, mirroring `_TARGETS` plus the
 * `require_current_revision(...)` mount sites in `app/services/revision_fence.py`
 * and the three routers that use it.
 *
 * `writes` is an ALLOWLIST of the sub-paths under `{root}/{id}` that the backend
 * actually fences (`''` being the record itself, i.e. `PUT`/`DELETE` on it). It is
 * not "every non-GET under a fenced record": the backend deliberately leaves
 * `/conversation/send-message` unfenced so a contact can still be messaged about a
 * revised record (AC O2), and `/view-link`, `/export/pdf`, `/sync-assignee` and
 * the manual template send are unfenced too. Stamping the header there is not
 * merely noise - it makes ANY 409 from those routes read as a revision conflict
 * and trigger a blanket cache invalidation for a conflict that has nothing to do
 * with revisions.
 *
 * An allowlist also fails safe the same way the backend does: a route added
 * upstream starts out unsent, and an absent header means unfenced, which is the
 * honest answer until this list learns about it.
 *
 * Complaint is listed even though its `portal_revision_configs` row ships
 * disabled: the backend target exists, so when the checkbox is flipped the
 * frontend arms itself with no code change (UAC CB4). Until the payload carries
 * a `revision_no` nothing is remembered and no header is sent.
 */
export const FENCED_COLLECTIONS = [
  {
    root: '/api/v1/procurement/stock-inquiries',
    writes: [
      '',
      'attachments',
      'response-attachments',
      'update-and-reply',
      'submit-for-project-sales',
      'project-sales-approve',
      'project-sales-reject',
      'purchasing-reject',
      'reopen',
      'void',
    ],
  },
  {
    root: '/api/v1/procurement/purchase-requests',
    writes: [
      '',
      'attachments',
      'update-and-reply',
      'set-pending-approval',
      'reject-submitted',
      'approval-decision',
      'process',
      'close',
      'void',
      'send-approval-link',
    ],
  },
  {
    root: '/api/v1/complaints-management/complaints',
    writes: [
      '',
      'attachments',
      'response-attachments',
      'update-and-reply',
      'approve',
      'reject',
      'process',
      'close',
      'void',
      'notify-root-cause',
      'notify-resolution',
    ],
  },
] as const;

type CollectionMatcher = {
  /** GET of the collection itself: the list the row actions are taken from. */
  list: RegExp;
  /** GET of one record: the detail page. */
  detail: RegExp;
  /** Any path under one record: `[1]` is the entity id, `[2]` the sub-path. */
  record: RegExp;
  /** The sub-paths the backend fences. `''` is the record itself. */
  writes: Set<string>;
};

const MATCHERS: CollectionMatcher[] = FENCED_COLLECTIONS.map(({ root, writes }) => ({
  list: new RegExp(`^${root}$`),
  detail: new RegExp(`^${root}/(${UUID_PATTERN})$`),
  record: new RegExp(`^${root}/(${UUID_PATTERN})(?:/(.*))?$`),
  writes: new Set<string>(writes),
}));

/**
 * The `/api/v1/...` path of a request, query string stripped.
 *
 * `apiFetch` may hand us an absolute URL (local dev points at :8000) or a
 * relative one (production goes through nginx), so match from `/api/v1/`
 * onwards rather than parsing an origin that may not be there.
 */
function apiPath(url: unknown): string | null {
  if (typeof url !== 'string') return null;
  const at = url.indexOf('/api/v1/');
  if (at < 0) return null;
  const path = url.slice(at);
  const cut = path.search(/[?#]/);
  return cut < 0 ? path : path.slice(0, cut);
}

/** True when this request is a GET whose body carries revisions worth keeping. */
export function isFencedReadPath(url: unknown, method?: string): boolean {
  if ((method ?? 'GET').toUpperCase() !== 'GET') return false;
  const path = apiPath(url);
  if (!path) return false;
  return MATCHERS.some((m) => m.list.test(path) || m.detail.test(path));
}

/**
 * The entity id a write is aimed at, when the URL is a route the backend fences.
 *
 * A sub-path the backend leaves unfenced returns null even though it sits under a
 * fenced record: no header goes out, and a 409 from it is left alone rather than
 * reported to the user as someone else's revision.
 */
export function fencedWriteEntityId(url: unknown, method?: string): string | null {
  const verb = (method ?? 'GET').toUpperCase();
  if (verb === 'GET' || verb === 'HEAD' || verb === 'OPTIONS') return null;
  const path = apiPath(url);
  if (!path) return null;
  for (const m of MATCHERS) {
    const hit = m.record.exec(path);
    if (!hit) continue;
    return m.writes.has(hit[2] ?? '') ? hit[1] : null;
  }
  return null;
}

// ---------------------------------------------------------------------------
// The registry: what revision each record was showing when we last read it.
// ---------------------------------------------------------------------------

const _seen = new Map<string, number>();

/** Remember one record's revision. Monotonic: never walks a number backwards. */
export function rememberRevision(id: unknown, revisionNo: unknown): void {
  if (typeof id !== 'string' || !id) return;
  // `Number(null)` is 0, and a payload with no `revision_no` at all - a type
  // whose column has not shipped yet - must read as UNKNOWN, not as revision 0.
  // Remembering 0 there would start sending a header for a record we never saw
  // a revision for.
  if (typeof revisionNo !== 'number' && typeof revisionNo !== 'string') return;
  if (typeof revisionNo === 'string' && !revisionNo.trim()) return;
  const next = Number(revisionNo);
  if (!Number.isFinite(next) || next < 0) return;
  const current = _seen.get(id);
  if (current !== undefined && current >= next) return;
  _seen.set(id, Math.trunc(next));
}

/** The revision we last saw for a record, or null when we never read it. */
export function rememberedRevision(id: string): number | null {
  const found = _seen.get(id);
  return found === undefined ? null : found;
}

/** Drop everything. Called on sign-out, and by tests between cases. */
export function clearRememberedRevisions(): void {
  _seen.clear();
}

/**
 * Harvest `{id, revision_no}` out of a fenced GET payload.
 *
 * Deliberately shallow: one record, or one page of them under `items` / `data`.
 * A deep walk would also swallow the revision TIMELINE payload, whose rows carry
 * a `revision_no` beside the revision row's own id, and poison the registry with
 * ids that are not entity ids.
 */
export function harvestRevisions(payload: unknown): void {
  if (!payload || typeof payload !== 'object') return;
  if (Array.isArray(payload)) {
    for (const row of payload) harvestRevisions(row);
    return;
  }
  const record = payload as Record<string, unknown>;
  const rows = record.items ?? record.data;
  if (Array.isArray(rows)) {
    for (const row of rows) {
      if (row && typeof row === 'object') {
        const entry = row as Record<string, unknown>;
        rememberRevision(entry.id, entry.revision_no);
      }
    }
    return;
  }
  rememberRevision(record.id, record.revision_no);
}

// ---------------------------------------------------------------------------
// Refusal handling: refresh the record, do not leave the user on a dead version.
// ---------------------------------------------------------------------------

type StaleHandler = (entityId: string) => void;

let _onStale: StaleHandler | null = null;

/**
 * Register what to do when the server refuses a write as stale.
 *
 * Wired once in `providers/query-provider.tsx` to invalidate the react-query
 * cache, which is what actually puts the new revision on screen. Kept as a
 * registration rather than importing a query client here so this module stays
 * free of React and usable from `lib/api.ts` on either side of the boundary.
 */
export function registerRevisionStaleHandler(handler: StaleHandler | null): void {
  _onStale = handler;
}

/**
 * Turn a fenced 409 into a refresh plus a response every caller reads correctly.
 *
 * The services in this repo read an error body three different ways
 * (`err.detail`, `err.message`, `extractApiError`). The server sends only
 * `detail`, so the two hand-rolled readers would have thrown `Error(undefined)`
 * and toasted nothing on exactly the conflict this feature exists to report.
 * Extract the sentence ONCE, here, with `extractApiError`, and hand back a 409
 * carrying it under both keys - so the sentence reaches the user through the
 * error path each call site already has, with no per-call-site change and no
 * second toast competing with the first.
 */
export async function handleRevisionConflict(
  entityId: string,
  response: Response,
): Promise<Response> {
  const message = await extractApiError(
    response.clone(),
    'This record was revised while you were working on it. Reload to see the latest revision.',
  );
  try {
    _onStale?.(entityId);
  } catch {
    /* a failed refresh must never swallow the refusal itself */
  }
  return new Response(JSON.stringify({ detail: message, message }), {
    status: 409,
    statusText: response.statusText,
    headers: { 'Content-Type': 'application/json' },
  });
}
